src/serac/physics/utilities/functional/domain_integral_cuda.cuh

```diff
@@ -0,0 +1,292 @@
+#pragma unroll
+  for (int qe = thread_id; qe < num_quadrature_points; qe += grid_stride) {
+    int e = qe / rule.size();
+    int q = qe % rule.size();
+    // get the (change in) values for this particular element
+    tensor du_elem = detail::Load<trial_element>(du, e);
+
+    // this is where we will accumulate the (change in) element residual tensor
+    element_residual_type dr_elem{};
+
+    gradient_quadrature<g, test, trial, Q, derivatives_type>(e, q, du_elem, dr_elem, derivatives_ptr, J, num_elements);
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    detail::Add(dr, dr_elem, e);
+  }
+}
+
+template <Geometry g, typename test, typename trial, int Q, serac::detail::ThreadExecutionPolicy policy,
+          typename derivatives_type>
+void gradient_kernel_cuda(serac::detail::ThreadExecutionConfiguration config, const mfem::Vector& dU, mfem::Vector& dR,
+                          derivatives_type* derivatives_ptr, const mfem::Vector& J_, int num_elements)
+{
+  using test_element               = finite_element<g, test>;
+  using trial_element              = finite_element<g, trial>;
+  using element_residual_type      = typename trial_element::residual_type;
+  static constexpr int  test_ndof  = test_element::ndof;
+  static constexpr int  trial_ndof = trial_element::ndof;
+  static constexpr auto rule       = GaussQuadratureRule<g, Q>();
+  static constexpr int  dim        = dimension_of(g);
+
+  // Use the device (GPU)
+  // J_.UseDevice(true);
+  // dU.UseDevice(true);
+  // dR.UseDevice(true);
+
+  // mfem provides this information in 1D arrays, so we reshape it
+  // into strided multidimensional arrays before using
+  auto J  = mfem::Reshape(J_.Read(), rule.size(), dim, dim, num_elements);
+  auto du = detail::Reshape<trial>(dU.Read(), trial_ndof, num_elements);
+  auto dr = detail::Reshape<test>(dR.ReadWrite(), test_ndof, num_elements);
+
+  cudaDeviceSynchronize();
+  serac::detail::displayLastCUDAErrorMessage(std::cout);
+
+  // call gradient_cuda
+  if constexpr (policy == serac::detail::ThreadExecutionPolicy::THREAD_PER_ELEMENT_QUADRATURE_POINT) {
+    int blocks_quadrature_element = (num_elements * rule.size() + config.blocksize - 1) / config.blocksize;
+    gradient_cuda_quadrature<g, test, trial, Q, derivatives_type>
+        <<<blocks_quadrature_element, config.blocksize>>>(du, dr, derivatives_ptr, J, num_elements);
+
+  } else if constexpr (policy == serac::detail::ThreadExecutionPolicy::THREAD_PER_ELEMENT) {
+    int blocks_element = (num_elements + config.blocksize - 1) / config.blocksize;
+    gradient_cuda_element<g, test, trial, Q, derivatives_type>
+        <<<blocks_element, config.blocksize>>>(du, dr, derivatives_ptr, J, num_elements);
+  }
+
+  cudaDeviceSynchronize();
+  serac::detail::displayLastCUDAErrorMessage(std::cout);
+  dR.HostRead();
+
+  // J_.UseDevice(false);
```

Comment:
dead code.

Reply:
Sorry, we're in the process of debugging this. So some of the comments may or may not come back.